### PR TITLE
Support creation of managed secret without associating it to a workload

### PR DESCRIPTION
**Pull request by DhritiShikhar**
Task: https://issues.redhat.com/browse/APPSVC-9

**Description**
Add support for "generic binding".
"In the service catalog UX this meant I just want the information dumped into a Secret, and then I'll choose what to do with it after that. This is good for referencing the Secret from other things like Jobs, or a Tekton pipeline, where your controller may not be able to directly modify a pod spec.
 
The request for this is coming up in discussions about transitioning some IBM brokers off of service catalog to operators instead. I think the use case is valid so we should consider it."

**Acceptance Criteria**
1. If the `ApplicationSelectors` is skipped, only the binding secret should be created with the name of the secret being dumped in the status for arbitrary consumption.
2. If the CR is updated with the `ApplicationSelectors` section, then the controller should inject the previously created secret into the deployment/deploymentconfig.
3. Irrespective of (2), If the credentials change on the backing service side, the binding secret should be updated as well

### CODE REVIEW

**Comment by sbose78** [comment]
I don't see code changes, only tests?

**Comment by DhritiShikhar** [comment]
@sbose78 All the 3 acceptance criteria is already supported. I wrote test to confirm the 3 acceptance criteria in the task:

1. If the `ApplicationSelectors` is skipped, only the binding secret should be created with the name of the secret being dumped in the status for arbitrary consumption.

2. If the CR is updated with the `ApplicationSelectors` section, then the controller should inject the previously created secret into the deployment/deploymentconfig.

3. Irrespective of (2), If the credentials change on the backing service side, the binding secret should be updated as well.

**Comment by sbose78** [comment]
Thank you!

**Comment by sbose78** [comment]
The API validations don't let it go through..

image

**Comment by sbose78** [comment]
.. but you would then get this

```
2020-02-26T11:29:56.327-0500    DEBUG   secret  Secret already exists, updating contents instead...     {"Namespace": "sbose", "Name": "binding-request"}
E0226 11:29:56.372848     560 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 914 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x2127980, 0x30586a0)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x82
panic(0x2127980, 0x30586a0)
        /usr/local/Cellar/go/1.13.8/libexec/src/runtime/panic.go:679 +0x1b2
github.com/redhat-developer/service-binding-operator/pkg/controller/servicebindingrequest.(*Binder).search(0xc00027bec0, 0x2, 0xc0004559a0, 0xf)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/pkg/controller/servicebindingrequest/binder.go:53 +0x7c
github.com/redhat-developer/service-binding-operator/pkg/controller/servicebindingrequest.(*Binder).Bind(0xc00027bec0, 0xc00088da28, 0x2, 0x2, 0xc0004559a0, 0xf)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/pkg/controller/servicebindingrequest/binder.go:539 +0x2f
github.com/redhat-developer/service-binding-operator/pkg/controller/servicebindingrequest.(*ServiceBinder).Bind(0xc0001b34a0, 0x232b02e, 0x30, 0x0, 0x0)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/pkg/controller/servicebindingrequest/binding.go:232 +0x179
github.com/redhat-developer/service-binding-operator/pkg/controller/servicebindingrequest.(*Reconciler).bind(0xc0003b2ba0, 0xc000780100, 0xc0001b34a0, 0xc00034a1b8, 0x0, 0x0, 0x0, 0x0)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/pkg/controller/servicebindingrequest/reconciler.go:93 +0xde
github.com/redhat-developer/service-binding-operator/pkg/controller/servicebindingrequest.(*Reconciler).Reconcile(0xc0003b2ba0, 0xc000b02fd9, 0x5, 0xc000b02fc0, 0xf, 0xc00088dcd8, 0xc0007485a0, 0xc000748518, 0x25002e0)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/pkg/controller/servicebindingrequest/reconciler.go:176 +0x5c2
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000366cc0, 0x217efe0, 0xc00038c380, 0x0)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:256 +0x162
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000366cc0, 0xc00012a700)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:232 +0xcb
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc000366cc0)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:211 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc00089a8d0)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00089a8d0, 0x3b9aca00, 0x0, 0xc000ab0501, 0xc0000be240)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc00089a8d0, 0x3b9aca00, 0xc0000be240)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:193 +0x328
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x1fb694c]

goroutine 914 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:55 +0x105
panic(0x2127980, 0x30586a0)
        /usr/local/Cellar/go/1.13.8/libexec/src/runtime/panic.go:679 +0x1b2
github.com/redhat-developer/service-binding-operator/pkg/controller/servicebindingrequest.(*Binder).search(0xc00027bec0, 0x2, 0xc0004559a0, 0xf)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/pkg/controller/servicebindingrequest/binder.go:53 +0x7c
github.com/redhat-developer/service-binding-operator/pkg/controller/servicebindingrequest.(*Binder).Bind(0xc00027bec0, 0xc00088da28, 0x2, 0x2, 0xc0004559a0, 0xf)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/pkg/controller/servicebindingrequest/binder.go:539 +0x2f
github.com/redhat-developer/service-binding-operator/pkg/controller/servicebindingrequest.(*ServiceBinder).Bind(0xc0001b34a0, 0x232b02e, 0x30, 0x0, 0x0)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/pkg/controller/servicebindingrequest/binding.go:232 +0x179
github.com/redhat-developer/service-binding-operator/pkg/controller/servicebindingrequest.(*Reconciler).bind(0xc0003b2ba0, 0xc000780100, 0xc0001b34a0, 0xc00034a1b8, 0x0, 0x0, 0x0, 0x0)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/pkg/controller/servicebindingrequest/reconciler.go:93 +0xde
github.com/redhat-developer/service-binding-operator/pkg/controller/servicebindingrequest.(*Reconciler).Reconcile(0xc0003b2ba0, 0xc000b02fd9, 0x5, 0xc000b02fc0, 0xf, 0xc00088dcd8, 0xc0007485a0, 0xc000748518, 0x25002e0)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/pkg/controller/servicebindingrequest/reconciler.go:176 +0x5c2
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000366cc0, 0x217efe0, 0xc00038c380, 0x0)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:256 +0x162
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000366cc0, 0xc00012a700)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:232 +0xcb
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc000366cc0)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:211 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc00089a8d0)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00089a8d0, 0x3b9aca00, 0x0, 0xc000ab0501, 0xc0000be240)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc00089a8d0, 0x3b9aca00, 0xc0000be240)
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
        /Users/sbose/appbinding/src/github.com/redhat-developer/service-binding-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:193 +0x328
```

**Comment by sbose78** [comment]
I tried making a PR to your branch but the history was a bit messed up ( my bad probably ), so created one against the main repo:
https://github.com/redhat-developer/service-binding-operator/pull/351 . I will not be merging this, we'll merge your PR ( current one ).

I've tested https://github.com/redhat-developer/service-binding-operator/pull/351, it works, but doesn't provide any feedback in the `status` block. 

* Could you please take in the changes, and then..
* make improvements if needed, 
* ensure status is updated
* add more tests based on the new changes

File: pkg/apis/apps/v1alpha1/zz_generated.openapi.go [hunk 1]
```diff
@@ -11,11 +11,11 @@ import (
...
-		"./pkg/apis/apps/v1alpha1.ServiceBindingRequestSpec":   schema_pkg_apis_apps_v1alpha1_ServiceBindingRequestSpec(ref),
-		"./pkg/apis/apps/v1alpha1.ServiceBindingRequestStatus": schema_pkg_apis_apps_v1alpha1_ServiceBindingRequestStatus(ref),
+		"github.com/redhat-developer/service-binding-operator/pkg/apis/apps/v1alpha1.ApplicationSelector":         schema_pkg_apis_apps_v1alpha1_ApplicationSelector(ref),
+		"github.com/redhat-developer/service-binding-operator/pkg/apis/apps/v1alpha1.BackingServiceSelector":      schema_pkg_apis_apps_v1alpha1_BackingServiceSelector(ref),
+		"github.com/redhat-developer/service-binding-operator/pkg/apis/apps/v1alpha1.ServiceBindingRequest":       schema_pkg_apis_apps_v1alpha1_ServiceBindingRequest(ref),
+		"github.com/redhat-developer/service-binding-operator/pkg/apis/apps/v1alpha1.ServiceBindingRequestSpec":   schema_pkg_apis_apps_v1alpha1_ServiceBindingRequestSpec(ref),
+		"github.com/redhat-developer/service-binding-operator/pkg/apis/apps/v1alpha1.ServiceBindingRequestStatus": schema_pkg_apis_apps_v1alpha1_ServiceBindingRequestStatus(ref),
```
Comment (isutton) [inline pkg/apis/apps/v1alpha1/zz_generated.openapi.go]: This is not the first time I've seen this; why generated files are being generated differently between developers?

Comment (sbose78) [reply pkg/apis/apps/v1alpha1/zz_generated.openapi.go]: Maybe because of operator-sdk versions? I'm using 0.5.2

Comment (DhritiShikhar) [reply pkg/apis/apps/v1alpha1/zz_generated.openapi.go]: Mine is different:

```
➜  service-binding-operator git:(generic) ✗ operator-sdk version                                                         
operator-sdk version: "v0.15.0-37-g0eac8a11", commit: "0eac8a1169dfc0a8918b9e41ada6278307a0852b", go version: "go1.13.6 linux/amd64"
```

**Comment by isutton** [comment]
> I tried making a PR to your branch but the history was a bit messed up ( my bad probably ), so created one against the main repo:
> #351 . I will not be merging #351, we'll merge your PR ( current one ).
> 
> I've tested #351, it works, but doesn't provide any feedback in the `status` block.
> 
>     * Could you please take in the changes, and then..
> 
>     * make improvements if needed,
> 
>     * ensure status is updated
> 
>     * add more tests based on the new changes
> 
> 
> Here's the only relevant commit there.
> [ef6f306](https://github.com/redhat-developer/service-binding-operator/commit/ef6f3066ee2d79e6cd4898da41aac35c3b760fbc)

Shouldn't we leave those changes for another PR, since the implemented tests already seem to validate the initial assumption that it currently works?

**Comment by sbose78** [comment]
> Shouldn't we leave those changes for another PR, since the implemented tests already seem to validate the initial assumption that it currently works?

Sorry, which changes are you saying we should skip in this PR ?

**Comment by sbose78** [comment]
How's this going? @DhritiShikhar

File: pkg/apis/apps/v1alpha1/zz_generated.openapi.go (hunk 1 above)
Comment (sbose78) [reply pkg/apis/apps/v1alpha1/zz_generated.openapi.go]: You definitely got to update your operator-sdk :-)

**Comment by isutton** [comment]
> > Shouldn't we leave those changes for another PR, since the implemented tests already seem to validate the initial assumption that it currently works?
> 
> Sorry, which changes are you saying we should skip in this PR ?

Related to the lack of feedback in the `status` block.

**Comment by sbose78** [comment]
But the `status` is empty.

**Comment by DhritiShikhar** [comment]
@sbose78 Hmm.. but I added check for `status` in my test. It's not empty there -> https://github.com/redhat-developer/service-binding-operator/pull/350/commits/572adc74db48e12ccf905b6b14dfb8633e93bb48

Both `status.BindingStatus` and `status.Conditions` are non-empty with desirable result.

**Comment by DhritiShikhar** [comment]
@Avni-Sharma addressed. Please review again f2c599a

File: pkg/controller/servicebindingrequest/reconciler_test.go [hunk 2]
```diff
@@ -205,3 +205,78 @@ func TestReconcilerReconcileUsingVolumes(t *testing.T) {
...
+
+	sbrOutput3, err := reconciler.getServiceBindingRequest(namespacedName)
+	require.NoError(t, err)
+
+	d = appsv1.Deployment{}
+	require.NoError(t, fakeClient.Get(ctx, namespacedName, &d))
+
+	require.Equal(t, "BindingSuccess", sbrOutput3.Status.BindingStatus)
+	require.Equal(t, corev1.ConditionTrue, sbrOutput3.Status.Conditions[0].Status)
+	require.Equal(t, reconcilerName, sbrOutput3.Status.Secret)
```
Comment (Avni-Sharma) [inline pkg/controller/servicebindingrequest/reconciler_test.go]: Can we have `sbrOutput3.Status.BindingData.Name` . Now that the bindingData can be either in a Secret or a ConfigMap according to #366 . Refer https://github.com/redhat-developer/service-binding-operator/pull/366/files#diff-14a37dadc390fc3ec23459efc088de8eR164-R180 cc @sbose78

Comment (DhritiShikhar) [reply pkg/controller/servicebindingrequest/reconciler_test.go]: but #366 is not merged yet

Comment (Avni-Sharma) [reply pkg/controller/servicebindingrequest/reconciler_test.go]: we can track this and send a new PR :)

File: pkg/controller/servicebindingrequest/reconciler_test.go [hunk 3]
```diff
@@ -205,3 +205,78 @@ func TestReconcilerReconcileUsingVolumes(t *testing.T) {
 		require.Equal(t, reconcilerName, volumes[0].VolumeSource.Secret.SecretName)
 	})
 }
+
+func TestReconcilerGenericBinding(t *testing.T) {
+	ctx := context.TODO()
+	backingServiceResourceRef := "test-using-volumes"
```
Comment (Avni-Sharma) [inline pkg/controller/servicebindingrequest/reconciler_test.go]: We are not using volumes for this test case so we should rename "test-using-volumes"

Comment (DhritiShikhar) [reply pkg/controller/servicebindingrequest/reconciler_test.go]: fixed 78118e8

File: pkg/apis/apps/v1alpha1/zz_generated.openapi.go (hunk 1 above)
Comment (Avni-Sharma) [reply pkg/apis/apps/v1alpha1/zz_generated.openapi.go]: I am on 0.15.2 and even for me these changes are happening

**Comment by DhritiShikhar** [comment]
@sbose78 Can you merge this PR (if no more comments) ?

**Comment by sbose78** [comment]
Could you share an example we've tested this with ? I'm getting the following:

```
apiVersion: apps.openshift.io/v1alpha1
kind: ServiceBindingRequest
metadata:
  creationTimestamp: '2020-03-12T21:05:44Z'
  generation: 1
  name: node-todo-git
  namespace: sbose
  resourceVersion: '221728'
  selfLink: >-
    /apis/apps.openshift.io/v1alpha1/namespaces/sbosse/servicebindingrequests/node-todo-git
  uid: 3413f95b-7cd6-44b7-9a62-daaec834be18
spec:
  backingServiceSelectors:
    - group: etcd.database.coreos.com
      kind: EtcdCluster
      resourceRef: example
      version: v1beta2
  detectBindingResources: true
status:
  bindingStatus: BindingFail
  conditions:
    - lastHeartbeatTime: '2020-03-12T21:05:59Z'
      lastTransitionTime: '2020-03-12T21:05:45Z'
      message: application ResourceRef or MatchLabel not found
      reason: BindingFail
      status: 'False'
      type: Ready
  secret: node-todo-git
```

**Comment by sbose78** [comment]
The secret does get created though.

**Comment by sbose78** [comment]
@DhritiShikhar Please use the issue template to specify "How to test". We should not be replacing the template with the content of a linked issue.

**Pull request merged**